File: backend-go/internal/services/providers/registry.go
package providers

import (
	"net/http"
	"sort"

	"github.com/siiway/voidswitch/internal/constants"
	"github.com/siiway/voidswitch/internal/database"
)

type ProviderInterface interface {
	BaseURL() string
	UpstreamURL() string
	ModelsURL() string
	GetBalanceURL() string
	Headers(apiKey string, extra map[string]string) map[string]string
	MapModel(model string) string
	PrepareBody(body map[string]any) map[string]any
	Classify(statusCode int, body any) ErrorClass
	FetchBalance(client *http.Client, apiKey string) (*BalanceResult, error)
	GetType() string
	GetStyle() constants.ApiStyle
	GetRecord() *database.Provider
}

type constructorFunc func(record *database.Provider) ProviderInterface

var constructors = map[string]constructorFunc{
	"openai":        func(r *database.Provider) ProviderInterface { return NewOpenAIProvider(r) },
	"openai-resp":   func(r *database.Provider) ProviderInterface { return NewOpenAIResponsesProvider(r) },
	"anthropic":     func(r *database.Provider) ProviderInterface { return NewAnthropicProvider(r) },
	"claude-code":   func(r *database.Provider) ProviderInterface { return NewClaudeCodeProvider(r) },
	"deepseek":      func(r *database.Provider) ProviderInterface { return NewDeepSeekProvider(r) },
	"siliconflow":   func(r *database.Provider) ProviderInterface { return NewSiliconFlowProvider(r) },
	"openrouter":    func(r *database.Provider) ProviderInterface { return NewOpenRouterProvider(r) },
	"groq":          func(r *database.Provider) ProviderInterface { return NewGroqProvider(r) },
	"xai":           func(r *database.Provider) ProviderInterface { return NewXAIProvider(r) },
	"moonshot":      func(r *database.Provider) ProviderInterface { return NewMoonshotProvider(r) },
	"mimo":          func(r *database.Provider) ProviderInterface { return NewMiMoProvider(r) },
	"nvidia":        func(r *database.Provider) ProviderInterface { return NewNvidiaProvider(r) },
	"mistral":       func(r *database.Provider) ProviderInterface { return NewMistralProvider(r) },
	"together":      func(r *database.Provider) ProviderInterface { return NewTogetherProvider(r) },
	"fireworks":     func(r *database.Provider) ProviderInterface { return NewFireworksProvider(r) },
	"perplexity":    func(r *database.Provider) ProviderInterface { return NewPerplexityProvider(r) },
	"cerebras":      func(r *database.Provider) ProviderInterface { return NewCerebrasProvider(r) },
	"cloudflare":    func(r *database.Provider) ProviderInterface { return NewCloudflareProvider(r) },
	"deepinfra":     func(r *database.Provider) ProviderInterface { return NewDeepInfraProvider(r) },
	"gemini":        func(r *database.Provider) ProviderInterface { return NewGeminiProvider(r) },
	"novita":        func(r *database.Provider) ProviderInterface { return NewNovitaProvider(r) },
	"sambanova":     func(r *database.Provider) ProviderInterface { return NewSambaNovaProvider(r) },
	"hyperbolic":    func(r *database.Provider) ProviderInterface { return NewHyperbolicProvider(r) },
	"nebius":        func(r *database.Provider) ProviderInterface { return NewNebiusProvider(r) },
	"github-models": func(r *database.Provider) ProviderInterface { return NewGitHubModelsProvider(r) },
	"zhipu":         func(r *database.Provider) ProviderInterface { return NewZhipuProvider(r) },
	"qwen":          func(r *database.Provider) ProviderInterface { return NewQwenProvider(r) },
	"volcengine":    func(r *database.Provider) ProviderInterface { return NewVolcengineProvider(r) },
	"minimax":       func(r *database.Provider) ProviderInterface { return NewMiniMaxProvider(r) },
	"generic":       func(r *database.Provider) ProviderInterface { return NewGenericOpenAIProvider(r) },
}

func AdapterTypes() []string {
	types := make([]string, 0, len(constructors))
	for t := range constructors {
		types = append(types, t)
	}
	sort.Strings(types)
	return types
}

func GetAdapter(record *database.Provider) ProviderInterface {
	c, ok := constructors[record.Type]
	if !ok {
		c = constructors["generic"]
	}
	return c(record)
}

func AdapterCatalog() []map[string]any {
	types := AdapterTypes()
	catalog := make([]map[string]any, 0, len(types))
	for _, t := range types {
		p := constructors[t](&database.Provider{Type: t})
		entry := map[string]any{
			"type":              p.GetType(),
			"style":             string(p.GetStyle()),
			"default_base_url":  p.BaseURL(),
			"default_models":    p.GetRecord().Models,
			"supports_balance":  p.GetBalanceURL() != "",
		}
		catalog = append(catalog, entry)
	}
	return catalog
}

var _ ProviderInterface = (*BaseProvider)(nil)
var _ ProviderInterface = (*OpenAIProvider)(nil)
var _ ProviderInterface = (*OpenAIResponsesProvider)(nil)
var _ ProviderInterface = (*AnthropicProvider)(nil)
var _ ProviderInterface = (*ClaudeCodeProvider)(nil)
var _ ProviderInterface = (*DeepSeekProvider)(nil)
var _ ProviderInterface = (*CloudflareProvider)(nil)
var _ ProviderInterface = (*GenericOpenAIProvider)(nil)

File: backend/voidswitch/services/providers/registry.py
"""Registry mapping a provider ``type`` string to its adapter class."""

from __future__ import annotations

from voidswitch.models.db import Provider

from .anthropic import AnthropicProvider
from .base import BaseProvider
from .cerebras import CerebrasProvider
from .claude_code import ClaudeCodeProvider
from .cloudflare import CloudflareProvider
from .deepinfra import DeepInfraProvider
from .deepseek import DeepSeekProvider
from .fireworks import FireworksProvider
from .gemini import GeminiProvider
from .generic import GenericOpenAIProvider
from .github_models import GitHubModelsProvider
from .grok import GrokProvider
from .groq import GroqProvider
from .hyperbolic import HyperbolicProvider
from .mimo import MiMoProvider
from .minimax import MiniMaxProvider
from .mistral import MistralProvider
from .moonshot import MoonshotProvider
from .nebius import NebiusProvider
from .novita import NovitaProvider
from .nvidia import NvidiaProvider
from .openai import OpenAIProvider
from .openai_responses import OpenAIResponsesProvider
from .openrouter import OpenRouterProvider
from .perplexity import PerplexityProvider
from .qwen import QwenProvider
from .sambanova import SambaNovaProvider
from .siliconflow import SiliconFlowProvider
from .together import TogetherProvider
from .volcengine import VolcengineProvider
from .xai import XAIProvider
from .zhipu import ZhipuProvider

_ADAPTERS: dict[str, type[BaseProvider]] = {
    cls.type: cls
    for cls in (
        OpenAIProvider,
        OpenAIResponsesProvider,
        AnthropicProvider,
        ClaudeCodeProvider,
        DeepSeekProvider,
        SiliconFlowProvider,
        OpenRouterProvider,
        GroqProvider,
        XAIProvider,
        GrokProvider,
        MoonshotProvider,
        MiMoProvider,
        NvidiaProvider,
        MistralProvider,
        TogetherProvider,
        FireworksProvider,
        PerplexityProvider,
        CerebrasProvider,
        CloudflareProvider,
        DeepInfraProvider,
        GeminiProvider,
        NovitaProvider,
        SambaNovaProvider,
        HyperbolicProvider,
        NebiusProvider,
        GitHubModelsProvider,
        ZhipuProvider,
        QwenProvider,
        VolcengineProvider,
        MiniMaxProvider,
        GenericOpenAIProvider,
    )
}


def adapter_types() -> list[str]:
    return sorted(_ADAPTERS)


def adapter_class(provider_type: str) -> type[BaseProvider]:
    return _ADAPTERS.get(provider_type, GenericOpenAIProvider)


def get_adapter(record: Provider) -> BaseProvider:
    return adapter_class(record.type)(record)


def adapter_catalog() -> list[dict[str, object]]:
    """Metadata used by the dashboard to populate the provider-type picker."""
    catalog: list[dict[str, object]] = []
    for name, cls in sorted(_ADAPTERS.items()):
        catalog.append(
            {
                "type": name,
                "style": cls.style.value,
                "default_base_url": cls.default_base_url,
                "default_models": list(cls.default_models),
                "supports_balance": cls.balance_suffix is not None,
                "supports_import": cls.supports_import,
            }
        )
    return catalog
